File: macos-app/Sources/ThoughtInput/Core/Destination.swift
import Foundation

// MARK: - Keychain Reference

struct KeychainRef: Codable, Sendable, Hashable {
    let account: String

    static func create() -> KeychainRef {
        KeychainRef(account: UUID().uuidString)
    }
}

// MARK: - Destination

struct Destination: Codable, Sendable, Identifiable {
    let id: UUID
    var name: String
    var isActive: Bool
    var type: DestinationType

    init(id: UUID = UUID(), name: String, isActive: Bool = false, type: DestinationType) {
        self.id = id
        self.name = name
        self.isActive = isActive
        self.type = type
    }
}

// MARK: - Destination Type

enum DestinationType: Codable, Sendable {
    case supabase(SupabaseConfig)
    case restNoAuth(RESTNoAuthConfig)
    case restApiKey(RESTApiKeyConfig)
    case restOAuthPassword(RESTOAuthPasswordConfig)
    case restOAuthClientCredentials(RESTOAuthClientCredentialsConfig)

    var displayName: String {
        switch self {
        case .supabase: "Supabase"
        case .restNoAuth: "REST (No Auth)"
        case .restApiKey: "REST (API Key)"
        case .restOAuthPassword: "REST (OAuth Password)"
        case .restOAuthClientCredentials: "REST (OAuth Client Credentials)"
        }
    }

    var iconName: String {
        switch self {
        case .supabase: "server.rack"
        case .restNoAuth: "network"
        case .restApiKey: "key"
        case .restOAuthPassword: "person.badge.key"
        case .restOAuthClientCredentials: "lock.shield"
        }
    }

    var keychainRefs: [KeychainRef] {
        switch self {
        case .supabase(let c): [c.apiKeyRef]
        case .restNoAuth: []
        case .restApiKey(let c): [c.apiKeyRef]
        case .restOAuthPassword(let c): [c.usernameRef, c.passwordRef]
        case .restOAuthClientCredentials(let c): [c.clientIDRef, c.clientSecretRef]
        }
    }
}

// MARK: - Config Structs

struct SupabaseConfig: Codable, Sendable {
    var projectURL: String
    var tableName: String
    var apiKeyRef: KeychainRef
}

struct RESTNoAuthConfig: Codable, Sendable {
    var endpointURL: String
}

struct RESTApiKeyConfig: Codable, Sendable {
    var endpointURL: String
    var headerName: String
    var apiKeyRef: KeychainRef
}

struct RESTOAuthPasswordConfig: Codable, Sendable {
    var endpointURL: String
    var tokenURL: String
    var usernameRef: KeychainRef
    var passwordRef: KeychainRef
}

struct RESTOAuthClientCredentialsConfig: Codable, Sendable {
    var endpointURL: String
    var tokenURL: String
    var clientIDRef: KeychainRef
    var clientSecretRef: KeychainRef
}

// MARK: - OAuth Token

struct OAuthToken: Codable, Sendable {
    let accessToken: String
    let expiresAt: Date?
    let refreshToken: String?

    var isExpired: Bool {
        guard let expiresAt else { return false }
        return Date.now.addingTimeInterval(30) >= expiresAt
    }
}

// MARK: - Pending Capture

struct PendingCapture: Codable, Sendable {
    let payload: CapturePayload
    let destinationID: UUID
    let destinationSnapshot: Destination
}

File: android-app/app/src/main/java/com/thoughtinput/capture/data/destinations/Destination.kt
package com.thoughtinput.capture.data.destinations

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import java.util.UUID

object UUIDSerializer : KSerializer<UUID> {
    override val descriptor: SerialDescriptor =
        PrimitiveSerialDescriptor("UUID", PrimitiveKind.STRING)

    override fun serialize(encoder: Encoder, value: UUID) {
        encoder.encodeString(value.toString())
    }

    override fun deserialize(decoder: Decoder): UUID = UUID.fromString(decoder.decodeString())
}

@Serializable
data class KeychainRef(val account: String) {
    companion object {
        fun create(): KeychainRef = KeychainRef(UUID.randomUUID().toString())
    }
}

@Serializable
data class Destination(
    @Serializable(with = UUIDSerializer::class) val id: UUID = UUID.randomUUID(),
    val name: String,
    val isActive: Boolean = false,
    val type: DestinationType
)

@Serializable
sealed class DestinationType {
    abstract val displayName: String

    @Serializable
    data class Supabase(
        val projectURL: String,
        val tableName: String,
        val apiKeyRef: KeychainRef
    ) : DestinationType() {
        override val displayName: String get() = "Supabase"
    }

    @Serializable
    data class RestNoAuth(
        val endpointURL: String
    ) : DestinationType() {
        override val displayName: String get() = "REST (No Auth)"
    }

    @Serializable
    data class RestApiKey(
        val endpointURL: String,
        val headerName: String,
        val apiKeyRef: KeychainRef
    ) : DestinationType() {
        override val displayName: String get() = "REST (API Key)"
    }

    @Serializable
    data class RestOAuthPassword(
        val endpointURL: String,
        val tokenURL: String,
        val usernameRef: KeychainRef,
        val passwordRef: KeychainRef
    ) : DestinationType() {
        override val displayName: String get() = "REST (OAuth Password)"
    }

    @Serializable
    data class RestOAuthClientCredentials(
        val endpointURL: String,
        val tokenURL: String,
        val clientIDRef: KeychainRef,
        val clientSecretRef: KeychainRef
    ) : DestinationType() {
        override val displayName: String get() = "REST (OAuth Client Credentials)"
    }

    val keychainRefs: List<KeychainRef>
        get() = when (this) {
            is Supabase -> listOf(apiKeyRef)
            is RestNoAuth -> emptyList()
            is RestApiKey -> listOf(apiKeyRef)
            is RestOAuthPassword -> listOf(usernameRef, passwordRef)
            is RestOAuthClientCredentials -> listOf(clientIDRef, clientSecretRef)
        }
}

data class OAuthToken(
    val accessToken: String,
    val expiresAt: Long?,
    val refreshToken: String?
) {
    val isExpired: Boolean
        get() = expiresAt?.let { System.currentTimeMillis() + 30_000L >= it } ?: false
}

@Serializable
data class PendingCapture(
    val payload: com.thoughtinput.capture.data.CapturePayload,
    @Serializable(with = UUIDSerializer::class) val destinationID: UUID,
    val destinationSnapshot: Destination
)
